client/client.go

```diff
@@ -717,13 +785,98 @@ func newWebsocketDialerTLS(conf Config) (*websocket.Dialer, error) {
 
 	dialer := websocket.Dialer{
 		NetDialTLSContext: func(_ context.Context, network string, addr string) (net.Conn, error) {
+			proxyUrl, err := getProxyUrlFromHostPort(addr)
+			if err != nil {
+				log.Errorf("Failed to get http-proxy configurations during websocket dial: %v", err)
+			} else if proxyUrl != nil {
+				return proxyDialOpenSSL(ctx, &conf, network, addr, proxyUrl)
+			}
 			return dialOpenSSL(ctx, &conf, network, addr)
 		},
 	}
 
 	return &dialer, nil
 }
 
+func getProxyUrlFromHostPort(addr string) (*url.URL, error) {
+	u, err := url.Parse("https://" + addr)
+	if err != nil {
+		return u, err
+	}
+	// presumably for security reasons, http.ProxyFromEnvironment returns
+	// nil if req.URL is localhost. To aid testing, allow localhost proxying
+	// only if both HTTPS_PROXY and req.URL are locahosts
+	if proxyUrl, err := url.Parse(os.Getenv("HTTPS_PROXY")); err == nil {
+		proxyHost := strings.Split(proxyUrl.Host, ":")[0]
+		targetHost := strings.Split(addr, ":")[0]
+		if targetHost == "127.0.0.1" && proxyHost == "127.0.0.1" {
+			return proxyUrl, err
+		}
+	}
+	return http.ProxyFromEnvironment(&http.Request{URL: u})
+}
+
+func getHostPort(u *url.URL) (hostPort string) {
+	hostPort = u.Host
+	if i := strings.LastIndex(u.Host, ":"); i > strings.LastIndex(u.Host, "]") {
+		return u.Host
+	} else {
+		switch u.Scheme {
+		case "https":
+			hostPort += ":443"
+		default:
+			hostPort += ":80"
+		}
+	}
+	return hostPort
+}
+
+type httpProxyDialer struct {
+	proxyURL *url.URL
+}
+
+func (hpd *httpProxyDialer) Dial(network string, addr string) (net.Conn, error) {
+	hostPort := getHostPort(hpd.proxyURL)
+	conn, err := net.Dial(network, hostPort)
+	if err != nil {
+		return nil, err
+	}
+
+	connectHeader := make(http.Header)
+	if user := hpd.proxyURL.User; user != nil {
+		proxyUser := user.Username()
+		if proxyPassword, passwordSet := user.Password(); passwordSet {
+			credential := base64.StdEncoding.EncodeToString([]byte(proxyUser + ":" + proxyPassword))
+			connectHeader.Set("Proxy-Authorization", "Basic "+credential)
+		}
+	}
+
+	connectReq := &http.Request{
+		Method: http.MethodConnect,
+		URL:    &url.URL{Opaque: addr},
+		Host:   addr,
+		Header: connectHeader,
+	}
+	if err := connectReq.Write(conn); err != nil {
+		conn.Close()
+		return nil, err
+	}
+	br := bufio.NewReader(conn)
+	resp, err := http.ReadResponse(br, connectReq)
+	if err != nil {
+		conn.Close()
+		return nil, err
+	}
+	defer resp.Body.Close()
+
+	if resp.StatusCode != 200 {
+		conn.Close()
+		f := strings.SplitN(resp.Status, " ", 2)
+		return nil, errors.New(f[1])
```

Comment:
What is this?

Reply:
Wow, this slipped my mind, but I didn't write this. Code taken from `gorilla/websocket`'s `proxy.go`.

Reply:
Ah perhaps I misread this comment. I thought u might have referred to possibility that an index-out-of-bounce error could occur. This line returns the an error with a value equal to the response text. But if the response doesn't have a propper readline like `HTTP/1.1 202 Accepted`, but `HTTP/1.1 202` (without the reason part) then this will cause a runtime error. 
I should also wrap this error giving info that this is in fact coming from the response itself.